flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+
+		// put some states
+		for (long namespace = 0;  namespace < 15; namespace++) {
+			for (int key = 0; key < 20; key++) {
+				String state = String.valueOf(namespace * key);
+				stateMap.put(key, namespace, state);
+				addToReferenceState(referenceStates, key, namespace, state);
+			}
+		}
+		verifyState(referenceStates, stateMap);
+
+		ThreadLocalRandom random = ThreadLocalRandom.current();
+		InternalKvState.StateIncrementalVisitor<Integer, Long, String> visitor =
+			stateMap.getStateIncrementalVisitor(5);
+		while (visitor.hasNext()) {
+			for (StateEntry<Integer, Long, String> stateEntry : visitor.nextEntries()) {
+				int key = stateEntry.getKey();
+				long namespace = stateEntry.getNamespace();
+				String state = stateEntry.getState();
+				assertEquals(state, stateMap.get(key, namespace));
+				int op = random.nextInt(3);
+				switch (op) {
+					case 0:
+						visitor.remove(stateEntry);
+						removeFromReferenceState(referenceStates, key, namespace);
+						break;
+					case 1:
+						String newState = state + "-update";
+						visitor.update(stateEntry, newState);
+						addToReferenceState(referenceStates, key, namespace, newState);
+						break;
+					default:
+						break;
+				}
+			}
+		}
+		verifyState(referenceStates, stateMap);
+
+		// validates that visitor will be invalid after state map is closed
+		InternalKvState.StateIncrementalVisitor<Integer, Long, String> closedVisitor =
+			stateMap.getStateIncrementalVisitor(5);
+		assertTrue(closedVisitor.hasNext());
+
+		stateMap.close();
+		assertFalse(closedVisitor.hasNext());
+	}
+
+	@Test
+	public void testPutAndGetNode() {
+		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
+		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(
+			keySerializer, namespaceSerializer, stateSerializer, spaceAllocator);
+		final int key = 10;
+		final long namespace = 0L;
+		final String valueString = "test";
+		SkipListKeySerializer<Integer, Long> skipListKeySerializer = new SkipListKeySerializer<>(keySerializer, namespaceSerializer);
+		SkipListValueSerializer<String> skipListValueSerializer = new SkipListValueSerializer<>(stateSerializer);
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		byte[] constructedKeyBytes = new byte[keyBytes.length + 1];
+		System.arraycopy(keyBytes, 0, constructedKeyBytes, 1, keyBytes.length);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(constructedKeyBytes);
+		int keyLen = keyBytes.length;
+		byte[] value = skipListValueSerializer.serialize(valueString);
+		stateMap.putNode(keyByteBuffer, 1, keyLen, value, false);
+		long node = stateMap.getNode(keyByteBuffer, 1, keyLen);
+		long valuePointer = SkipListUtils.helpGetValuePointer(node, spaceAllocator);
+		Assert.assertThat(stateMap.helpGetState(valuePointer), is(valueString));
```

Comment:
This test seems to test only internal methods of the `CopyOnWriteSkipListStateMap`. This becomes also clear by looking at the code to generate the right internal format. I think we should not test these implementation details. Ideally we would test `put` and `get` by using the public API of `CopyOnWriteSkipListStateMap`.

Reply:
`CopyOnWriteSkipListStateMap.putNode` is a fundamental function and now implemented in a general way to accept all non-zero `keyOffset` but now all production codes only invoke the method with zero offset. The test here is trying to guard the correctness of the non-zero case to make sure we're not committing some "only-work-with-zero-param" function. Let me modify the test case name to reflect the purpose.

Reply:
We don't need non-zero offset for now, but may do further improvements like using partial `ByteBuffer` based on it. Requiring the offset to be zero is a hard limit, and since the correctness is already assured through unit test case, I'd say reserving it has no harm?

Reply:
Let's keep the code. In the future I would refrain from adding code which won't be needed in the immediate future because the past has shown that it might take quite a while until new features are implemented which actually needed these kind of functionality. Until then one needs to maintain the code without gaining more functionality.